_posts/tutorials/2021-05-20-frontends-with-javalinmithril.md

```diff
@@ -0,0 +1,450 @@
+Let's add it to `app-frame.js`:
+
+```javascript
+/*
+ @package io.javalin.mithril.demo;
+ */
+class AppFrame {
+    constructor(content) {
+    }
+
+    view(vnode) {
+        return m("div.app-frame", [
+            m("header", [
+                m("span", "Javalin Mithril Demo App"),
+                    window.javalin.state.currentUser ? m("span", `Current User : '${window.javalin.state.currentUser}'`) : null,
+                ]),
+            vnode.attrs.content
+            ]
+        )
+    }
+}
+
+```
+
+## Conclusion
+We've created a fully working (but pretty limited) admin interface with only a few files.
+* `App.java` contains the server config (routes, error handlers, access management)
+* `UserController.java` contains the list of fake users, and methods to get them (getAll, getOne)
+* `layout.html` loads the frontend dependencies and initializes Mithril
+* `app-frame.js` has a header and some global styling which is included in all components
+* `user-overview.js` displays a list of users
+* `user-profile.js` displays additional details for one user (requires login)
+* `not-found.js` displays a 404 error page
+* `user-model.js` contains the ViewModel for users
+* `pom.xml` contains all our dependencies
+
+Since our frontend dependencies are prepacked WebJars, we don't need NPM, and we don't need to
+manage any frontend libraries manually. The project structure is very clean:
+
+<div class="compressed-code" markdown="1">
+```kotlin
+javalinvue-example
+├───src
+│   └─── main
+│       └───java
+│       |   ├───io/javalinmithril/demo
+│       |        └───'Java Source Code'
+│       └───resources/mihtril
+│               └───app-frame.js
+│               └───not-found.js
+│               └───user-overview.js
+│               └───user-profile.js
+|               └───user-model.js
+│               └───layout.html
+└───pom.xml
+```
+<style>.compressed-code .highlighter-rouge pre code { line-height: 1.2; }</style>
+</div>
+
+That's about it. The Afterword contains a bit more discussion about this technique,
+but similarly to the Background section you can skip it if you're not that interested. **Thanks for reading**!
+
+### Afterword
+The Library is still in its early days. However, there is a lot of potential here for growing front-ends without the associated tool-chain complexities, keeping a small overhead/request - and keeping that overhead consistent, even as your application grows in features and complexity. While Mithril.js is designed around single-page applications, it just has so much our of the box that when we were considering the hyperscript framework to use for this plugin(Trust me when I say, it was even more bonkers than it is right now. Who even writes an import and package system?), we just decided to go with it because of the (outwardly looking) reactive nature, performance, and built in XHR functionality. 
```

Comment:
```suggestion
The Library is still in its early days. However, there is a lot of potential here for growing front-ends without the associated tool-chain complexities, keeping a small overhead/request - and keeping that overhead consistent, even as your application grows in features and complexity. While Mithril.js is designed around single-page applications, it just offers so much out of the box that when we were considering the hyperscript framework to use for this plugin (trust me when I say, it was even more bonkers than it is right now - who even writes an import and package system?), we just decided to go with it because of the (outwardly looking) reactive nature, performance, and built in XHR functionality. 
```